### ISSUE TITLE
[Question] Mime types and favicon.ico

### ISSUE THREAD

**Issue by a1730** [activej/activej#36]
We are testing ActiveJ with some of our users and we're getting 404s and wrong content-type on some files but favicon.ico is particularly sturborn.  We are trying to have the same look, feel and bugs with our current application.

We added a StaticServlet to handle static content, that was when we noticed that the artifacts were served with a content-type of  application/octet-stream. We tried to use io.activej.http.ContentTypes but the types are incomplete for our use case and, unfortunately, the class is final.

Therefore, starting with a copy of io.activej.http.ContentTypes and armed with ideas from [java2s](http://www.java2s.com/Code/Java/Network-Protocol/enumMimeType.htm), we created a contentTypeResolver attaching it toStaticServlet withContentTypeResolver. We started getting the right content-type **but** our resolver is not called for favicon.ico no matter what we do even though other files (css, js, and fonts) seem to be processed properly.

Neither,
`RoutingServlet.create()
      .map(GET, "/", StaticServlet.ofClassPath(executor, "static/").withIndexHtml())
      .map(GET, "/favicon.ico", StaticServlet.ofClassPath(executor, "static/favicon.ico").withContentTypeResolver(...) ...`
nor
`RoutingServlet.create()
      .map(GET, "/", StaticServlet.ofClassPath(executor, "static/").withContentTypeResolver().withIndexHtml() ...)
      .map(GET, "/favicon.ico", StaticServlet.ofClassPath(executor, "static/favicon.ico")` is working

> Junior developer question
Do we have to do anything special with favicon.ico? We are getting the right content-type and 200 with other files (css, js, fonts, etc.) but always 404 with favicon.ico.

Thank you.

**Comment by eduard-vasinskyi** [comment]
Hello, @a1730 

We don't quite understand the problem with 'favicon.ico'. For `StaticServlet` it is an ordinary file, same as any other file, so if you are able to serve other files, you should be able to serve 'favicon.ico' as well. In branch 'master-3.1' we have added a favicon to `StaticServletExample`. If you run the example, you should see that favicon is served successfully. If this does not solve your problem, could you please send a snippet of code that reproduces a bug? With index.html and resources structure?

As for using a custom type resolver, we have added an ability to statically register your own media types for custom extensions. You can do it by calling `MediaTypes::register`. 
Like this: `MediaTypes.register("example/custom-type", ".cstm")`
This change was introduced in commit 9d9d6505 and will be part of the ActiveJ v3.1 release.

**Comment by a1730** [comment]
Hello @eduard-vasinskyi,

I am glad to report that the favicon problem was **not** an activej issue. It was due to differences in browser behavior.

We found that Firefox (where we noticed the issue) and Chrome (no issues) differ in how they handle favicon.ico when there is a simple default-src CSP (Content-Security-Policy) on a site. The issue disappeared on Firefox after we added an explicit image-src csp.

Regarding type resolver, we serve many atypical file types and we do not trust browser sniff. Furthermore, we would like to add types by updating a simple mime-types.properties file without a requirement to rebuild and redeploy the application. Therefore, we have built a custom type resolver based on [concurrent-trees](https://github.com/npgall/concurrent-trees)  with an added plus of flexible searches - content-type, filename ends-with, etc. 

Thank you for a great project - We ❤️ activej here.
